### PR TITLE
[bc-breaking][quant][be] Add `is_qat` as a required argument to eager and fx graph apis

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #70009

Summary:
Currently we rely on module.training to decide whether we'll do a qat fusion or ptq fusion, this is
not ideal since training flag has nothing to do with quantization, this PR introduces an extra flag `is_qat`
to control this

We added fuse_modules_qat for fusions for qat and changed fuse_modules for fusions for PTQ.

This added is_qat (as last positional argument) to:
Eager:
fuse_known_modules,
fuser_methods in fuser_method_mapping

FX Graph:
[non-public] prepare function in torch.ao.quantization.fx.prepare

Note: currently we still has the constraint that when `is_qat` is True, the modules must be in training mode, we
can relax this constraint later

BC-breaking note:
Previous:
```
import torch
from torch.ao.quantization import fuse_modules
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = torch.nn.Conv2d(3, 3, 3)
        self.bn = torch.nn.BatchNorm2d(3)

    def forward(self, x):
        return self.bn(self.conv(x))

m = M().train()
m = fuse_modules(m, ["conv", "bn"])
print(type(m.conv))
m = M().eval()
m = fuse_modules(m, ["conv", "bn"])
print(type(m.conv))
```
Result:
```
<class 'torch.nn.intrinsic.modules.fused.ConvBn2d'>
<class 'torch.nn.modules.conv.Conv2d'>
```

After:
```
import torch
from torch.ao.quantization import fuse_modules
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = torch.nn.Conv2d(3, 3, 3)
        self.bn = torch.nn.BatchNorm2d(3)

    def forward(self, x):
        return self.bn(self.conv(x))

m = M().train()
m = fuse_modules_qat(m, ["conv", "bn"])
print(type(m.conv))
m = M().eval()
m = fuse_modules(m, ["conv", "bn"])
print(type(m.conv))
```

Result (doesn't change):
```
<class 'torch.nn.intrinsic.modules.fused.ConvBn2d'>
<class 'torch.nn.modules.conv.Conv2d'>
```

Test Plan:
```
python test/test_quantization.py TestFuseFx
python test/test_quantization.py TestFuseEager
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33178977](https://our.internmc.facebook.com/intern/diff/D33178977)

### CODE REVIEW

**Comment by vkuzo** [comment]
> Note: currently we still has the constraint that when is_qat is True, the modules must be in training mode, we
can relax this constraint later

would be good to do this in the same release as this PR

**Review by vkuzo — APPROVED** [review]

**Comment by jerryzh168** [comment]
> > Note: currently we still has the constraint that when is_qat is True, the modules must be in training mode, we
> > can relax this constraint later
> 
> would be good to do this in the same release as this PR

sure, that sounds good

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33178977).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33178977).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33178977).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33178977).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33178977).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33178977).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33178977).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33178977).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33178977).

**Comment by jerryzh168** [comment]
recreated in https://github.com/pytorch/pytorch/pull/70622 since github token expires

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33178977).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33178977).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33178977).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33178977).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33178977).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33178977).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33178977).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33178977).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33178977).

**Comment by malfet** [comment]
Reverting as it broke quick lints, see https://github.com/pytorch/pytorch/runs/4959313864?check_suite_focus=true (and same signal on PR)

**Comment by jerryzh168** [comment]
See GitHub Actions build Lint / quick-checks (2/2)
Step: "Ensure no trailing spaces" (full log | diagnosis details | 🔁 rerun)

2022-01-26T06:15:43.0179709Z python: can't open..._launches.py': [Errno 2] No such file or directory
2022-01-26T06:15:42.9851459Z ##[group]Run set -eux
2022-01-26T06:15:42.9851687Z �[36;1mset -eux�[0m
2022-01-26T06:15:42.9852012Z �[36;1mpython torch/testing/_check_kernel_launches.py |& tee "${GITHUB_WORKSPACE}"/cuda_kernel_launch_checks.txt�[0m
2022-01-26T06:15:42.9901393Z shell: /bin/bash -e {0}
2022-01-26T06:15:42.9901599Z env:
2022-01-26T06:15:42.9901846Z   pythonLocation: /opt/hostedtoolcache/Python/3.10.2/x64
2022-01-26T06:15:42.9902168Z   LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.10.2/x64/lib
2022-01-26T06:15:42.9902420Z ##[endgroup]
2022-01-26T06:15:42.9978596Z + tee /home/runner/work/pytorch/pytorch/cuda_kernel_launch_checks.txt
2022-01-26T06:15:43.0011221Z + python torch/testing/_check_kernel_launches.py
2022-01-26T06:15:43.0179709Z python: can't open file '/home/runner/work/pytorch/pytorch/torch/testing/_check_kernel_launches.py': [Errno 2] No such file or directory
2022-01-26T06:15:43.0234251Z ##[group]Run (! git --no-pager grep -I -no $'#include <cub/' --  ./aten  ':(exclude)aten/src/ATen/cuda/cub*.cuh' || (echo "The above files have direct cub include; please include ATen/cuda/cub.cuh instead and wrap your cub calls in at::native namespace if necessary"; false))
2022-01-26T06:15:43.0235078Z �[36;1m(! git --no-pager grep -I -no $'#include <cub/' --  ./aten  ':(exclude)aten/src/ATen/cuda/cub*.cuh' || (echo "The above files have direct cub include; please include ATen/cuda/cub.cuh instead and wrap your cub calls in at::native namespace if necessary"; false))�[0m
2022-01-26T06:15:43.0275617Z shell: /bin/bash -e {0}
2022-01-26T06:15:43.0276015Z env:
2022-01-26T06:15:43.0276328Z   pythonLocation: /opt/hostedtoolcache/Python/3.10.2/x64
2022-01-26T06:15:43.0276663Z   LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.10.2/x64/lib
2022-01-26T06:15:43.0276974Z ##[endgroup]
2022-01-26T06:15:43.0659433Z ##[group]Run (! git --no-pager grep -I -no $'cudaStreamSynchronize' --  ./aten ./c10 ':(exclude)aten/src/ATen/test' ':(exclude)c10/cuda/CUDAFunctions.h' || (echo "The above files call raw cuda APIs directly; please use at::cuda wrappers instead"; false))
2022-01-26T06:15:43.0660217Z �[36;1m(! git --no-pager grep -I -no $'cudaStreamSynchronize' --  ./aten ./c10 ':(exclude)aten/src/ATen/test' ':(exclude)c10/cuda/CUDAFunctions.h' || (echo "The above files call raw cuda APIs directly; please use at::cuda wrappers instead"; false))�[0m
2022-01-26T06:15:43.0702072Z shell: /bin/bash -e {0}

This really looked like a transient error that is not related to the PR, could you improve the message?

Or could we just block landing any PRs with lint errors?
